core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java

```diff
@@ -401,6 +401,23 @@ private void buildUberJar0(CurateOutcomeBuildItem curateOutcomeBuildItem,
             copyCommonContent(runnerZipFs, concatenatedEntries, applicationArchivesBuildItem, transformedClasses,
                     generatedClasses,
                     generatedResources, seen, finalIgnoredEntries);
+            // now that all entries have been added, check if there's a META-INF/versions/ entry. If present,
+            // mark this jar as multi-release jar. Strictly speaking, the jar spec expects META-INF/versions/N
+            // directory where N is an integer greater than 8, but we don't do that level of checks here but that
+            // should be OK.
+            if (Files.isDirectory(runnerZipFs.getPath("META-INF", "versions"))) {
+                log.debug("uber jar will be marked as multi-release jar");
+                final Path manifestPath = runnerZipFs.getPath("META-INF", "MANIFEST.MF");
+                final Manifest manifest = new Manifest();
+                // read the existing one
+                try (final InputStream is = Files.newInputStream(manifestPath)) {
+                    manifest.read(is);
+                }
+                manifest.getMainAttributes().put(Attributes.Name.MULTI_RELEASE, "true");
```

Comment:
Better test if `manifestPath` exists first?

Reply:
We know it exists because it is generated, however there is a comment at the generation saying it needs to be the first entry in the jar. 

Is there any problems with just marking all of them as multi release jars in the generateManifest code?

Reply:
> Is there any problems with just marking all of them as multi release jars in the generateManifest code?

That's an interesting point. I will experiment a bit to see if it has potential to cause issues.

Reply:
That may have a negative impact on performance. There is a reason that Oracle has made multi-release JAR authors include the flag in manifests instead of just checking for the presence of `META-INF/versions`. The manifest is probably already available, whereas checking for folders inside JAR files may require scanning the entire contents.

For this PR that check should be sufficient, because otherwise you'd have to go through the manifest files of all bundled dependencies, as well as that of the application itself. The chances of using `META-INF/versions` without wanting a multi-release JAR is very, very small.

Reply:
@stuartwdouglas, I had a look at the JDK code (especially the `JarFile`). Marking all (uber) jars as `Multi-Release`, as far as I can see won't have a functional impact. However, it does have a performance penalty. The code in `java.util.jar.JarFile` upon finding that a jar is multi-release will go into a code path which has additional `synchronized` blocks. When it sees that there are no versioned entries, it then fallbacks to the code that represents a non mutil-release jar. This additional synchronization penalty will be applicable every time there's a `JarEntry` being fetched (which can be high).

So I think the current approach in the PR to mark it as multi-release only when we find the `META-INF/versions`, IMO, would be a better approach. Does that sound fine?